packages/common/src/hooks/chats/useAudiusLinkResolver.ts

```diff
@@ -0,0 +1,210 @@
+ *      links in the provided value. This should be used before text submission.
+ *
+ * - getMatches: Function that returns a generator of matches for resolved
+ *      Audius links within the provided value. This is useful when displaying text
+ *      back to the user.
+ *
+ * - handleBackspace: Function that checks if the user backspaced over a resolved link
+ *      and modifies the text accordingly, deleting the whole display text/link on a
+ *      single keystroke.
+ *
+ * - reset: Function that resets the `trackId` and `collectionId` state to null.
+ *      Call this method when submitting.
+ */
+export const useAudiusLinkResolver = ({
+  value,
+  hostname,
+  audiusSdk
+}: {
+  value: string
+  hostname: string
+  audiusSdk: () => Promise<AudiusSdk>
+}) => {
+  // Maintain bidirectional maps of audius links to human readable format
+  const linkToHuman = useRef<{ [key: string]: string }>({}).current
+  const humanToData = useRef<{
+    [key: string]: { link: string; data: Track | Playlist | User }
+  }>({}).current
+
+  // The track id used to render the composer preview
+  const [trackId, setTrackId] = useState<Nullable<ID>>(null)
+  // The collection id used to render the composer preview
+  const [collectionId, setCollectionId] = useState<Nullable<ID>>(null)
+
+  /**
+   * Resolves Audius links in the provided text to their display text equivalent.
+   */
+  const resolveLinks = useCallback(
+    async (value: string) => {
+      const sdk = await audiusSdk()
+      const { matches } = matchAudiusLinks({
+        text: value,
+        hostname
+      })
+
+      for (const match of matches) {
+        if (!(match in linkToHuman)) {
+          const res = await sdk.resolve({ url: match })
+          if (res.data) {
+            if (instanceOfTrackResponse(res)) {
+              const human = formatTrackName({ track: res.data })
+              linkToHuman[match] = human
+              humanToData[human] = { link: match, data: res.data }
+            } else if (instanceOfPlaylistResponse(res)) {
+              const human = formatCollectionName({ collection: res.data[0] })
+              linkToHuman[match] = human
+              humanToData[human] = { link: match, data: res.data[0] }
+            } else if (instanceOfUserResponse(res)) {
+              const human = formatUserName({ user: res.data })
+              linkToHuman[match] = human
+              humanToData[human] = { link: match, data: res.data }
+            }
+          }
+        } else {
+          // If we already loaded the track, delay showing by 500ms
+          // to give the user some sense of what is happening.
+          await new Promise((resolve) => setTimeout(resolve, 500))
+        }
+      }
+
+      // Sort here to make sure that we replace all content before
+      // replacing their substrings.
+      let editedValue = value
+      for (const [link, human] of Object.entries(linkToHuman).sort(
```

Comment:
not sure i understand why sorting is necessary..

Reply:
we want to replace 

https://dharit/track

before we replace

https://dharit

because the latter is a substring of the former

Reply:
whoa neat